drivers/test.go

```diff
@@ -330,6 +335,69 @@ func (td *TestDriver) applyMessageExpectCodeAndReturn(msg *types.Message, code e
 	return result
 }
 
+//
+// Signed Message Appliers
+//
+
+func (td *TestDriver) ApplyMessageSigned(msg *types.Message) (result chain.ApplyResult) {
+	defer func() {
+		if r := recover(); r != nil {
+			result.Receipt.ExitCode = exitcode.SysErrInternal
+			td.T.Fatalf("message application panicked: %v", r)
+		}
+	}()
+	serMsg, err := msg.Serialize()
+	require.NoError(td.T, err)
+
+	msgSig, err := td.Wallet().Sign(msg.From, serMsg)
+	require.NoError(td.T, err)
+
+	smgs := &types.SignedMessage{
+		Message:   *msg,
+		Signature: msgSig,
+	}
+	result, err = td.validator.ApplySignedMessage(td.ExeCtx, td.State(), smgs)
+	require.NoError(td.T, err)
+	return result
+}
+
+func (td *TestDriver) ApplySignedOk(msg *types.Message) chain.ApplyResult {
+	return td.ApplySignedExpect(msg, EmptyReturnValue)
+}
+
+func (td *TestDriver) ApplySignedExpect(msg *types.Message, retval []byte) chain.ApplyResult {
+	return td.applyMessageSignedExpectCodeAndReturn(msg, exitcode.Ok, retval)
+}
+
+func (td *TestDriver) ApplySignedFailure(msg *types.Message, code exitcode.ExitCode) chain.ApplyResult {
+	return td.applyMessageExpectCodeAndReturn(msg, code, EmptyReturnValue)
+}
+
+func (td *TestDriver) applyMessageSignedExpectCodeAndReturn(msg *types.Message, code exitcode.ExitCode, retval []byte) chain.ApplyResult {
+	result := td.ApplyMessageSigned(msg)
+
+	td.GasMeter.Track(result.Receipt)
+
+	if td.Config.ValidateExitCode() {
+		assert.Equal(td.T, code, result.Receipt.ExitCode, "Expected ExitCode: %s Actual ExitCode: %s", code.Error(), result.Receipt.ExitCode.Error())
+	}
+	if td.Config.ValidateReturnValue() {
+		assert.Equal(td.T, retval, result.Receipt.ReturnValue, "Expected ReturnValue: %v Actual ReturnValue: %v", retval, result.Receipt.ReturnValue)
+	}
+	if td.Config.ValidateGas() {
+		expectedGasUsed, ok := td.GasMeter.NextExpectedGas()
+		if ok {
+			assert.Equal(td.T, expectedGasUsed, result.Receipt.GasUsed, "Expected GasUsed: %d Actual GasUsed: %d", expectedGasUsed, result.Receipt.GasUsed)
+		} else {
+			td.T.Logf("WARNING: failed to find expected gas cost for message: %+v", msg)
```

Comment:
This code is duplicated because we need need `msg` for a helpful debug message, could remove?

Reply:
Yea prefer to de-dupe. You can use an `interface{}` for the message debugging?

Reply:
how about returning a boolean value, false if gas isn't found?